### PR TITLE
Remember default currencies on a per-account basis rather than globally.

### PR DESCRIPTION
**Pull request by pmarciniak**
Global default currency doesn't work well for ledgers that have accounts denominated in various currencies, for 2 main reasons:

1. It will always use the currency of the very first transaction in the ledger as a default. For people who have moved countries, this will always be wrong, as it won't be the currency of the current country.
2. When interleaving accounts denominated in different currencies while entering new transactions, one has to constantly change the currency from the default being the last transaction's currency.

### CODE REVIEW

File: bean-add [hunk 1]
```diff
@@ -329,12 +330,12 @@ def read_account():
 		if reading == '':
 			return None
 		# Add new account option
-		if reading not in data.accounts:
+		if reading not in data.accounts_currencies.keys():
```
[the remark this answers is not in the document]
Comment (pmarciniak) [reply bean-add]: Done

File: bean-add [hunk 2]
```diff
@@ -1500,9 +1510,9 @@ readline.set_completer(complete)
 
 # Print some statistics
 print('\n%s processed' % pluralize(len(data.journal), 'record'))
-if len(data.accounts) > 0:
-	print('%s, %s, %s and %s loaded' % (
-		pluralize(len(data.accounts), 'account name'),
+if len(data.accounts_currencies.keys()) > 0:
```
[the remark this answers is not in the document]
Comment (pmarciniak) [reply bean-add]: Done

File: bean-add [hunk 3]
```diff
@@ -402,19 +405,23 @@ def read_amount(account):
...
 	set_history_context('amount')
 	# Read an amount
-	if account in data.accounts:
-		account = '`%s`' % account
-	amount = input('Enter the amount for %s (including currency symbol) [%s %s]: ' % (account, undecimal(data.balance), data.cur))
-	amount = [data.balance, data.cur] if amount == '' else amount.split()
+	default_currency = data.cur
+	if account in data.accounts_currencies.keys() and data.accounts_currencies[account] is not None:
+		default_currency = data.accounts_currencies[account]
+	account_printed = '`%s`' % account if account in data.accounts_currencies.keys() else account
```
[the remark this answers is not in the document]
Comment (pmarciniak) [reply bean-add]: Done

File: bean-add [hunk 4]
```diff
@@ -641,7 +648,7 @@ def tx_lookup(patterns, prune_txids=[]):
 # Find transactions that contain a string
 def find_transactions(autoseek=None):
 	# Use all possible completion contenxts for lookups
-	data.vocab = data.accounts + list(data.defaults.keys()) + ['#' + x for x in data.tags]
+	data.vocab = data.accounts_currencies.keys() + list(data.defaults.keys()) + ['#' + x for x in data.tags]
```
[the remark this answers is not in the document]
Comment (pmarciniak) [reply bean-add]: Done

File: bean-add [hunk 5]
```diff
@@ -37,7 +37,8 @@ class data(object):
 	prices = {}
 	quote_currency = ''
 	# Completion
-	accounts = []
+	# A mapping from known accounts to their respective default currencies.
+	accounts_currencies = {}
```
[the remark this answers is not in the document]
Comment (pmarciniak) [reply bean-add]: Done

**Review by simon-v — APPROVED** [review]

**Comment by simon-v** [comment]
> One can tell that I'm not used to writing python code...

Don't worry about it. I often find myself looking at my own code, wondering what i was thinking when i wrote it.

There is a new issue this PR introduced, though:

With currency defaults pulled from the mapping, transactions with different currencies and no conversion are allowed through without a warning. For example:
```
2023-05-13 * "Payment for item"
  Expenses:Goods 100 USD
  Assets:Cash -100 EUR
```
I'm currently researching possible solutions, but suggestions are welcome.

**Comment by simon-v** [comment]
Partially resolved in ecc1540, but more testing is needed.

**Pull request merged**